src/ext/detective.lisp

```diff
@@ -0,0 +1,282 @@
+                  (cons (cons "packages" package-regex) :package-reference)
+                  (cons (cons "variables" variable-regex) :variable-reference)
+                  (cons (cons "misc" misc-regex) :misc-reference))))
+    
+    (setf (buffer-references (current-buffer))
+          (make-hash-table :test 'equal))
+
+      (loop :for ((id . regex) . class) :in slots
+            :when (and regex (capture-regex-function regex))
+            :do (setf (gethash id (buffer-references (current-buffer)))
+                      (find-ref regex class)))))))
+
+(defgeneric capture-reference (position class))
+
+(defun %get-reference (references)
+  (alexandria:when-let* ((name-references (mapcar #'reference-name references))
+                         (item
+                          (prompt-for-string "Navigate: "
+                                             :completion-function (lambda (x) (completion-strings x name-references))
+
+                                             :test-function (lambda (name)
+                                                              (member name name-references :test #'string=)))))
+    (find item references :key #'reference-name :test #'string=)))
+
+(defgeneric navigate-reference (references))
+
+(defmethod navigate-reference ((type String))
+  (alexandria:when-let ((references (gethash type (buffer-references (current-buffer)))))
+    (%get-reference references)))
+
+(defmethod navigate-reference ((references List))
+  (%get-reference references))
+
+(defgeneric move-to-reference (reference))
+
+(defmethod move-to-reference ((reference reference))
+  (let ((location (reference-point reference)))
+    (move-point (current-point) location)))
+
+(defmethod move-to-reference (reference)
+  (message "Not reference available in current buffer."))
+
+(defun check-change ()
+  (cond 
+   ((null (buffer-references (current-buffer)))
+    (search-references 
+     (variable-value 'lem/language-mode:detective-search :buffer)))
+   
+   ((changed-disk-p (current-buffer))
+    (search-references (variable-value 'lem/language-mode:detective-search :buffer)))))
+
+(define-command detective-next () ()
+  (check-change)
+  (let* ((references (buffer-references (current-buffer)))
+         (lreferences (alexandria:hash-table-values references))
+         closest )
+    (loop :for ref :in (alexandria:flatten lreferences)
+          :for flag = t then nil
+          :with cline = (+ (line-number-at-point (current-point)) 1)
+          :for rline = (line-number-at-point (reference-point ref))
+          :do (progn
+                (when flag (setf closest ref))
+                (let* ((cdiff (- (line-number-at-point (reference-point closest)) cline))
+                       (ldiff (- rline cline)))
+
+                  (when  (or (and (> rline cline)
+                                  (<  ldiff cdiff))
+                             (< cdiff 0))
+                    (setf closest ref)))))
```

Comment:
There is a function called `lem:point<`.
For example, could it be written as follows?

```common-lisp
(loop :for (start end) :on ...points...
      :do (when (point< start (current-point) end)
            (return end)))

```

Reply:
So, I added a function to the point files to compare a list of points relative to the desired point and it returns the closest one down or up with optionally check if the point is in the same line :+1:  [(src/base/point.lisp)](https://github.com/lem-project/lem/pull/788/files#diff-6f18753392faf0e63518ba0d628865c3844a1a9dd9752befe8dbd9d94dc3c7afR196)

Reply:
This allowed me to remove most of the code from the next and previous functions, let me know what do you think